### PR TITLE
adding client certificate and key

### PR DESCRIPTION
**Pull request by raoel**
hey there,

for our case we needed to add a client certificate and key to the configuration - I'd like to upstream my changes.

I hope this passes your standards - I've modified existing unit tests for configuration (I didn't do a full integration test with certificates because that would make the testing setup a lot more complex)

### CODE REVIEW

**Comment by codebien** [comment]
Hi @raoel,
thanks for your contribution. :bow: 

Can you sign the CLA, please? We need it to accept the contribution.

**Comment by raoel** [comment]
> Hi @raoel, thanks for your contribution. bow
> 
> Can you sign the CLA, please? We need it to accept the contribution.

had to get permission from my employer :-) should be fixed.

**Comment by codebien** [comment]
Hi @raoel,
can you better explain your use case and how you're configuring the remote write, please?

Are you using a self-signed certificate or an internal CA? You should not share the pair with the client but just provide the CA file. https://prometheus.io/docs/guides/tls-encryption/#testing

**Comment by raoel** [comment]
hello (sorry for the delay... things go slow - it's been christmas here too ;-)  )

We have mutual TLS, so instead of providing the client with a username/password we need to provide the client with it's key and certificate to be able to authenticate. So, as far as I know, the client really need the pair.

File: README.md [hunk 1]
```diff
@@ -32,6 +32,15 @@ K6_PROMETHEUS_RW_PASSWORD=bar \
...
 
+or with client certificate:
+```
+K6_PROMETHEUS_RW_SERVER_URL=https://localhost:9090/api/v1/write \
+K6_PROMETHEUS_RW_INSECURE_SKIP_TLS_VERIFY=false \
+K6_PROMETHEUS_RW_CLIENT_CERTIFICATE=client.crt \
+K6_PROMETHEUS_RW_CLIENT_CERTIFICATE_KEY=client.key \
+./k6 run -o xk6-prometheus-rw script.js
+```
+
```
Comment (codebien) [inline README.md:43]: We should document that in the case a different CA is required then the relative operating systems' storage shouid be used for providing the custom CA.

File: pkg/remotewrite/config.go [hunk 2]
```diff
@@ -89,6 +97,16 @@ func (conf Config) RemoteConfig() (*remote.HTTPConfig, error) {
 		InsecureSkipVerify: conf.InsecureSkipTLSVerify.Bool, //nolint:gosec
 	}
 
+	if conf.ClientCertificate.Valid && conf.ClientCertificateKey.Valid {
```
Comment (codebien) [inline pkg/remotewrite/config.go:100]: I would prefer if we skip the assignment if Insecure is set to `true`.

File: pkg/remotewrite/config.go [hunk 3]
```diff
@@ -89,6 +97,16 @@ func (conf Config) RemoteConfig() (*remote.HTTPConfig, error) {
...
+		cert, _ := tls.LoadX509KeyPair(conf.ClientCertificate.String, conf.ClientCertificateKey.String)
+		hc.TLSConfig.Certificates = []tls.Certificate{cert}
+	}
+
+	if conf.ClientCertificate.Valid && conf.ClientCertificateKey.Valid {
+		cert, _ := tls.LoadX509KeyPair(conf.ClientCertificate.String, conf.ClientCertificateKey.String)
+		hc.TLSConfig.Certificates = []tls.Certificate{cert}
+	}
```
Comment (codebien) [inline pkg/remotewrite/config.go:108]: Are these duplicated lines?

File: pkg/remotewrite/config.go [hunk 4]
```diff
@@ -139,6 +157,22 @@ func (conf Config) Apply(applied Config) Config {
...
+		conf.ClientCertificateKey = applied.ClientCertificateKey
+	}
+
+	if applied.ClientCertificate.Valid {
+		conf.ClientCertificate = applied.ClientCertificate
+	}
+
+	if applied.ClientCertificateKey.Valid {
+		conf.ClientCertificateKey = applied.ClientCertificateKey
+	}
```
Comment (codebien) [inline pkg/remotewrite/config.go:174]: Are they duplicated?

File: pkg/remotewrite/config_test.go [hunk 5]
```diff
@@ -393,6 +393,74 @@ func TestOptionTrendAsNativeHistogram(t *testing.T) {
...
+	t.Parallel()
+
+	cases := map[string]struct {
+		arg     string
+		env     map[string]string
+		jsonRaw json.RawMessage
+	}{
+		"JSON": {jsonRaw: json.RawMessage(`{"clientCertificate":"client.crt","clientCertificateKey":"client.key"}`)},
+		"Env":  {env: map[string]string{"K6_PROMETHEUS_RW_CLIENT_CERTIFICATE": "client.crt", "K6_PROMETHEUS_RW_CLIENT_CERTIFICATE_KEY": "client.key"}},
+		//"Arg":  {arg: "username=user1,password=pass1"},
```
Comment (codebien) [inline pkg/remotewrite/config_test.go:406]:
```suggestion
```

File: pkg/remotewrite/config_test.go [hunk 6]
```diff
@@ -393,6 +393,74 @@ func TestOptionTrendAsNativeHistogram(t *testing.T) {
...
+		t.Run(name, func(t *testing.T) {
+			c, err := GetConsolidatedConfig(
+				tc.jsonRaw, tc.env, tc.arg)
+			require.NoError(t, err)
+			assert.Equal(t, expconfig, c)
+		})
+	}
+}
+
+func TestOptionClientCertificate(t *testing.T) {
```
Comment (codebien) [inline pkg/remotewrite/config_test.go:430]: Duplicated?

**Review by codebien — CHANGES_REQUESTED** [review]
Hey @raoel, sorry for the late reply :pray: Thanks again for your contribution. I've take a quick look and I've added some comments.

I didn't test yet it on a real setup, I will be on PTO for the rest of the week. I will do more tests on Monday.

**Review by codebien — APPROVED** [review]
Hey @raoel, thanks again for the contribution. I merge this PR on a development branch where we will do some polish of the feature then we will merge on the main branch. I will make sure to preserve you as the author/contributor of your commits.

**Comment by codebien** [comment]
Unfortunately, this branch contains conflicts and we can't push here because the PR is generated from a fork. Going to continue the work on #143 preserving the commit history and the original author.

**Pull request closed without merging**